aptos-move/move-examples/shared_account/sources/SharedAccount.move

```diff
@@ -0,0 +1,195 @@
+    const EINSUFFICIENT_BALANCE: u64 = 5;
+
+    fun find_greatest_common_divisor(denominator1: u64, denominator2: u64) : u64 {
+        if (denominator1 == 0) {
+            denominator2
+        }
+        else {
+            find_greatest_common_divisor(denominator2 % denominator1, denominator1)
+        }
+    }
+
+    fun find_least_common_denominator(denominator1: u64, denominator2: u64) : u64 {
+        assert!(denominator1 * denominator2 > 0, Errors::invalid_argument(EINVALID_INPUT));
+        let gcd: u64 = find_greatest_common_divisor(denominator1, denominator2);
+        assert!(gcd > 0, Errors::invalid_argument(Errors::invalid_argument(EINVALID_INPUT)));
+        (denominator1 * denominator2) / gcd
+    }
+
+    // Create and initialize a shared account
+    public fun initialize(source: &signer, seed: vector<u8>, addresses: vector<address>, numerators: vector<u64>, denominators: vector<u64>): address acquires SharedAccount {
+        assert!(Vector::length(&addresses) == Vector::length(&numerators), Errors::invalid_argument(EINVALID_INPUT));
+        assert!(Vector::length(&addresses) == Vector::length(&denominators), Errors::invalid_argument(EINVALID_INPUT));
+
+        assert!(Account::exists_at(Signer::address_of(source)), Errors::invalid_argument(EACCOUNT_NOT_FOUND));
+        let (resource_signer, resource_signer_cap) = Account::create_resource_account(source, seed);
+        if (!exists<SharedAccount>(Signer::address_of(&resource_signer))) {
+            move_to(
+                &resource_signer,
+                SharedAccount {
+                    share_record: Vector::empty<State>(),
+                    signer_capability: resource_signer_cap,
+                }
+            )
+        };
+
+        let shared_account = borrow_global_mut<SharedAccount>(Signer::address_of(&resource_signer));
+
+        let i = 0;
+        let cumulative_numerator = 0;
+        let cumulative_denominator = 1;
+
+        while (i < Vector::length(&addresses)) {
+            let num = *Vector::borrow(&numerators, i);
+            let denom = *Vector::borrow(&denominators, i);
+            let addr = *Vector::borrow(&addresses, i);
+            assert!(Account::exists_at(addr), Errors::invalid_argument(EACCOUNT_NOT_FOUND));
+            assert!(num >= 0 && denom > 0, Errors::invalid_argument(EINVALID_INPUT));
+
+            let common_denom = find_least_common_denominator(cumulative_denominator, denom);
+            cumulative_numerator = cumulative_numerator * (common_denom / cumulative_denominator) + num * (common_denom / denom);
+            cumulative_denominator = common_denom;
+
+            Vector::push_back(&mut shared_account.share_record, State { share_holder: addr, numerator: num, denominator: denom });
+            i = i + 1;
+        };
+
+        assert!(cumulative_numerator == cumulative_denominator, Errors::invalid_argument(EINVALID_NUMERATOR_DENOMINATOR_COMBINATIONS));
+        Signer::address_of(&resource_signer)
```

Comment:
probably good to emit event
and if we really need this, make it a private function and have the public version entry pointed

Reply:
event!